pkg/graveler/joined_diff_iterator.go

```diff
@@ -0,0 +1,116 @@
+package graveler
+
+import "bytes"
+
+// JoinedDiffIterator calculate the union diff between 2 iterators.
+// The output iterator yields a single result for each key.
+// If a key exist in the 2 iterators, iteratorA value prevails.
+type JoinedDiffIterator struct {
+	iterA        DiffIterator
+	iterAHasMore bool
+	iterB        DiffIterator
+	iterBHasMore bool
+	p            DiffIterator
+	started      bool
+}
+
+func NewJoinedDiffIterator(iterA DiffIterator, iterB DiffIterator) *JoinedDiffIterator {
+	return &JoinedDiffIterator{
+		iterA:        iterA,
+		iterAHasMore: true,
+		iterB:        iterB,
+		iterBHasMore: true,
+		p:            nil,
+	}
+}
+
+func (c *JoinedDiffIterator) Next() bool {
+	valA := c.iterA.Value()
+	valB := c.iterB.Value()
+	switch {
+	case !c.started:
+		// first
+		c.iterAHasMore = c.iterA.Next()
+		c.iterBHasMore = c.iterB.Next()
+		c.started = true
+	case !c.iterAHasMore && !c.iterBHasMore:
+		// last
+		return false
+	case !c.iterAHasMore:
+		// iterA is done
+		c.p = c.iterB
+		c.iterBHasMore = c.iterB.Next()
+	case !c.iterBHasMore:
+		// iterB is done
+		c.p = c.iterA
+		c.iterAHasMore = c.iterA.Next()
+	case bytes.Equal(valA.Key, valB.Key):
+		c.iterAHasMore = c.iterA.Next()
+		c.iterBHasMore = c.iterB.Next()
+	case bytes.Compare(valA.Key, valB.Key) < 0:
+		c.iterAHasMore = c.iterA.Next()
+		c.iterBHasMore = true
+	default:
+		// value of iterA < value of iterB
+		c.iterBHasMore = c.iterB.Next()
+		c.iterAHasMore = true
+	}
+	if c.iterA.Err() != nil {
+		c.p = c.iterA
+		c.iterAHasMore = false
+		c.iterBHasMore = false
+		return false
+	}
+	if c.iterB.Err() != nil {
+		c.p = c.iterB
+		c.iterAHasMore = false
+		c.iterBHasMore = false
+		return false
+	}
+	if !c.iterAHasMore && !c.iterBHasMore {
+		return false
```

Comment:
I don't understand.  Suppose that the last values of iterA and iterB refer to the same key.  Then [these lines](https://github.com/treeverse/lakeFS/pull/7701/files#diff-cc5982b5afba401bcedb16fb0c275f35653384ee6bcda09d4590a0e58131bf80R47-R49) will make this condition true -- and the iterator misses the last value!

Reply:
We process only values we already returned in the first switch-case statement.
I'll add a test case just to make sure.